runtime/near-vm-runner/src/combined_runner.rs

```diff
@@ -0,0 +1,471 @@
+#[cfg(feature = "combined_vm")]
+pub mod combined_runner {
+
+    use near_runtime_fees::RuntimeFeesConfig;
+    use near_vm_errors::HostError;
+    use near_vm_errors::{InconsistentStateError, VMError};
+    use near_vm_logic::types::{AccountId, Balance, Gas, ProfileData, PromiseResult, PublicKey};
+    use near_vm_logic::{External, VMConfig, VMContext, VMLogicError, VMOutcome, ValuePtr};
+    use std::collections::HashMap;
+
+    /// Proxy postponing the actual DB access, reads go via an actual external, writes are postponed
+    /// and accumulated.
+    pub struct ProxyExternal<'a> {
+        ext: &'a mut dyn External,
+        local_trie: HashMap<Vec<u8>, Vec<u8>>,
+        receipts: Vec<Receipt>,
+        validators: HashMap<AccountId, Balance>,
+        log: Vec<Action>,
+    }
+
+    impl<'a> ProxyExternal<'a> {
+        pub fn new(proxied: &'a mut dyn External) -> Self {
+            Self {
+                ext: proxied,
+                local_trie: HashMap::new(),
+                receipts: Vec::new(),
+                validators: HashMap::new(),
+                log: Vec::new(),
+            }
+        }
+
+        fn commit_log(&'a self) -> Vec<Action> {
+            self.log.clone()
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub enum Action {
+        CreateAccount(CreateAccountAction),
+        CreateReceipt(Receipt),
+        StorageSet(StorageSetAction),
+        StorageRemove(StorageRemoveAction),
+        DeployContract(DeployContractAction),
+        FunctionCall(FunctionCallAction),
+        Transfer(TransferAction),
+        Stake(StakeAction),
+        AddKeyWithFullAccess(AddKeyWithFullAccessAction),
+        AddKeyWithFunctionCall(AddKeyWithFunctionCallAction),
+        DeleteKey(DeleteKeyAction),
+        DeleteAccount(DeleteAccountAction),
+    }
+
```

Comment:
I don't think we should do this. It's too much replication and we should refactor primitives instead. Otherwise it'll become unmaintainable.

Reply:
Well, here we implement one feature (running with two VMs), if primitives has to be refactored, it's better be done separately. Overall enumeration of all `External` trait operations is needed, and how it is achieved is an implementation detail. My question was mostly about semantics of operation, and if postponing of side effects is implemented in the manner compatible with the transaction runtime.

Reply:
I am confused. Why couldn't we do it the same way as `RuntimeExt`?